src/backend/api/meals.js

```diff
@@ -2,14 +2,202 @@ const express = require("express");
+
+router.put("/:id", async (req, res) => {
+  try {
+    const updatedMeal = req.body;
+    const mealId = req.params.id;
+    const meal = await knex("meal").where("id", mealId).update(updatedMeal);
+    if (meal) {
+      res.status(200).json({
+        message: `Meal with id - ${mealId} was updated`,
+      });
+    } else {
+      res.status(404).send("Meal doesn't exist");
+      return;
+    }
+  } catch (error) {
+    res.status(500).json("Server Error");
+  }
+});
+
+// DELETE - /api/meals/:id - deletes the meal by id
+// routes should handle the case when the row with that ID does not exist
+
+router.delete("/:id", async (req, res) => {
+  try {
+    const mealId = req.params.id;
+    const deletedMeal = await knex("meal").where("id", mealId).del();
+    if (deletedMeal) {
+      res.status(200).send("Meal is successfully deleted");
+    } else {
+      res.status(404).send("Meal doesn't exist");
+      return;
+    }
+  } catch (error) {
+    res.status(500).json("Server Error");
     throw error;
   }
 });
 
+// nodejs-week3
+
+// GET - /api/meals/:meal_id/reviews - returns all reviews for a specific meal
+
+router.get("/:id/reviews", async (req, res) => {
+  try {
+    const mealId = req.params.id;
+    const mealReviews = await knex("meal")
+      .join("review", "review.meal_id", "=", "meal.id")
+      .select("meal.id", "meal.title", "review.description")
+      .where({ "review.meal_id": mealId });
+    res.json(mealReviews);
+  } catch (error) {
+    res.status(500).send("Server Error");
+  }
+});
+
+// nodejs-week3 - adding query parameters
+
+router.get("/", async (req, res) => {
+  try {
+    const {
+      maxPrice,
+      availableReservations,
+      title,
+      dateAfter,
+      dateBefore,
+      limit,
+      sortKey,
+      sortDir,
+    } = req.query;
+    let mealsQuery = knex("meal");
+
+    // GET - api/meals?maxPrice=number - returns all meals that are cheaper than maxPrice
+    //parameter - maxPrice, data type - number
+
+    if (maxPrice) {
+      mealsQuery = mealsQuery.where("price", "<=", maxPrice);
+    }
+
+    // GET - api/meals?availableReservations=true - returns all meals that still have available spots left
+    //parameter - availableReservations, data type - boolean
+
+    if (availableReservations === "true") {
+      mealsQuery = mealsQuery
+        .innerJoin("reservation", "meal.id", "reservation.meal_id")
+        .groupBy("meal.id")
+        .select("meal.*")
+        .having(
+          "meal.max_reservations",
+          ">",
+          knex.sum("reservation.number_of_guests")
+        );
+    } else if (availableReservations === "false") {
+      mealsQuery = mealsQuery
+        .innerJoin("reservation", "meal.id", "reservation.meal_id")
+        .groupBy("meal.id")
+        .select("meal.*")
+        .having(
+          "meal.max_reservations",
+          "<=",
+          knex.sum("reservation.number_of_guests")
+        );
+    }
```

Comment:
Should we perhaps do something special if they have passed something that is not `true` or `false` to `availableReservations`? Can you think of a fitting HTTP status code? 😄